### PR TITLE
Implemented Constrained Planar Cuts algorithm (CVPR 2015)

### PR DESCRIPTION
**Pull request by mschoeler**
Corrected some formating mistakes with LCCP
Extended LCCP code to serve as a base for CPC.

Paper:
http://www.cv-foundation.org/openaccess/content_cvpr_2015/papers/Schoeler_Constrained_Planar_Cuts_2015_CVPR_paper.pdf

### CODE REVIEW

**Comment by taketwo** [comment]
Is there any reason why `WeightedRandomSampleConsensus` is embedded into the new segmentation class? Does it make sense to contribute it separately to the `sample_consensus` module?

**Comment by mschoeler** [comment]
Actually I had it as a separate class in the sample consesus module before. I included it into the CPC class,  because I think the weighted RANSAC is too specific to the problem I had to solve:

- I use normals as the direction of the edges in my adjacency graph. So the implementation of weightsac uses this for the directional scoring.
- Weightsac only works with plane models, because I can easily get the plane-normals for the directed cuts. Using for example a sphere model one would need to calculate normals depending on the location of the model.
- I normalize my score by the total number of inliers in the plane. So the score (for undirected cuts) effectively becomes the fraction of concave connections vs. the total connections cut (concave points have a weight of 1 and convex points a weight of 0).
This is counter intuitive to other RANSAC algorithms, because there you try to maximize the number of inliers. Short example: If you set all weights to the same number, the algorithm would give all planes the same score.

So I think it would cause more confusion, if I create a public interface and made it available to other users.

**Comment by taketwo** [comment]
Ok, thanks for the explanation, makes sense. But I think I need to read the paper anyway :)
Would it be possible for you to extract the changes to LCCP into a separate pull request? We'll review and merge it first, and then move on with the new algorithm.

**Comment by mschoeler** [comment]
sure I will do this!

**Comment by mschoeler** [comment]
ok, I split of the changes to LCCP into this pull request  #1287

**Comment by jspricke** [comment]
@taketwo I guess we can merge this, right?

**Comment by taketwo** [comment]
Doesn't compile, @mschoeler can you have a look?

File: segmentation/include/pcl/segmentation/cpc_segmentation.h [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    public:
+      CPCSegmentation ();
+      virtual
+      ~CPCSegmentation ();
+
+      /** \brief Merge supervoxels using local convexity. The input parameters are generated by using the SupervoxelClustering class. To retrieve the output use the relabelCloud method (inherited from LCCPSegmentation).
+       *  \param[in] supervoxel_clusters_arg Map of < supervoxel labels, supervoxels >
+       *  \param[in] label_adjacency_arg The graph defining the supervoxel adjacency relations  */
+      void
+      segment (std::map<uint32_t, typename pcl::Supervoxel<PointT>::Ptr> &supervoxel_clusters_arg,
```
Comment (taketwo) [inline segmentation/include/pcl/segmentation/cpc_segmentation.h]: If these are "in" parameters as the docs suggest, then they can be made `const`.

**Comment by mschoeler** [comment]
Thanks for the comments. I'm on it. Strangely GCC compiles just fine; clang does not. 
About the const. I guess I already overlooked it in LCCP. This function has the same interface.
You once said you do not like the interface of LCCP (and CPC) too much and I agree. Maybe I should make a setInputSupervoxels (which takes the supervoxels and adjacency as argument). Segment would than return the labeled PointCloud (what relabel cloud is doing right now). I guess than this would be more consistent to f.e. the euclidean clustering class. What do you think?

**Comment by taketwo** [comment]
+1 to have a `setInputSupervoxels`, this would be in line with the rest of PCL, where classes are stateful. Regarding the output of the `segment()` function, however, I think more consistent would be to output a vector of vectors of supervoxel labels. This is similar to how other segmentation classes output a vector of vector of point indices.

```cpp
void segment(std::vector<std::vector<uint32_t> >& segments);
```

**Comment by mschoeler** [comment]
`void segment(std::vector<std::vector<uint32_t> >& segments)` is a good idea. However I think it should output the indices of the points in the pointcloud, not the supervoxels.
Also I think that we need a method which is able to apply `std::vector<std::vector<uint32_t> > segments` onto a pointclouds label field. Maybe a functions exists already??
For example: `labelPointCloud<Intype, Outtype> ( const pcl::PointCloud<Intype> &in_cloud, std::vector<std::vector<uint32_t> > segments, pcl::PointCloud<Outtype> &out_cloud)`.
That could be used whenever one wants to save the output of clustering algorithms.

**Comment by mschoeler** [comment]
@taketwo  what is your opinion on that?

**Comment by taketwo** [comment]
> However I think it should output the indices of the points in the pointcloud, not the supervoxels.

But the input is a bunch of supervoxels, is the algorithm even aware that originally there was some underlying point cloud?

But I see your point. I think there could be uses for both kinds of output.

> Also I think that we need a method which is able to apply `std::vector<std::vector<uint32_t> >` segments onto a pointclouds label field. Maybe a functions exists already?

I'm not aware of such function, but I definitely see a lot of use for it.

**Comment by mschoeler** [comment]
You are right and I does not make sense to give lccp a pointer to the cloud.
After some experimenting, I think `segment` should not return data.
Instead the user can decide which kind of data he/she wants to get.
There are three methods for that:
- `getSupervoxelToSegmentMap`: Get a map between the supervoxel-id and the segment
- `getSegmentToSupervoxelMap`: Retrieving all supervoxels within a segment (this is close to what you wanted for the `segment` function. I does not return a `vector <vector <uint32_t> >` because that would require copying data. Instead it returns `std::map<uint32_t, std::set<uint32_t> >`.
But I could do that, too. I do not feel strong about that.
- `relabelCloud`: As before. Relabels the supervoxel-label field by the segment. 

So not much changed. I added a method `setInputSupervoxels`, though.
I will create a pull request for LCCP and you'll see what I mean.

**Comment by mschoeler** [comment]
So I created a pull request for the changes I talked about #1389.

**Comment by mschoeler** [comment]
any other comments on this one? I'm now incorporating the changes from the interface (same as LCCP) and see what caused the compiler error.

File: segmentation/include/pcl/segmentation/cpc_segmentation.h [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+#include <pcl/point_types.h>
+#include <pcl/point_cloud.h>
+
+// segmentation and sample consensus includes
+#include <pcl/segmentation/supervoxel_clustering.h>
+#include <pcl/segmentation/lccp_segmentation.h>
+#include <pcl/sample_consensus/sac.h>
+
+//pcl::PointXYZINormal is not a core point type (so we cannot use the precompiled classes here)
+#ifndef PCL_ONLY_CORE_POINT_TYPES
```
Comment (mschoeler) [inline segmentation/include/pcl/segmentation/cpc_segmentation.h]: @taketwo The problem was that PointXYZINormal is not a core point type. This is why the functions (extract_clusters and sample_consensus classes)  have not been compiled for this point type and I got the linker error.
Is this a good way to deal with this? What do you think?

**Comment by mschoeler** [comment]
@taketwo I cannot figure out why travis does not build for the third test case.
I cannot reproduce the compiler error on my machine when I run the same settings:
`PCL_ONLY_CORE_POINT_TYPES=ON` and `PCL_NO_PRECOMPILE=ON` `gcc`


Here is the error I get with Travis:
```
/home/travis/build/PointCloudLibrary/pcl/segmentation/include/pcl/segmentation/impl/cpc_segmentation.hpp: In member function ‘void pcl::CPCSegmentation<PointT>::applyCuttingPlane(uint32_t)’:
/home/travis/build/PointCloudLibrary/pcl/segmentation/include/pcl/segmentation/impl/cpc_segmentation.hpp:90:48: error: type/value mismatch at argument 1 in template parameter list for ‘template<class _Tp, class _Alloc> class std::vector’
/home/travis/build/PointCloudLibrary/pcl/segmentation/include/pcl/segmentation/impl/cpc_segmentation.hpp:90:48: error:   expected a type, got ‘EdgeID’
```

EdgeID is a typedef from LCCP which I included into CPC with the `using typename LCCP::EdgeID` directive (file `cpc_segmentation.h line 74+`):

```
  template <typename PointT>
  class CPCSegmentation : public LCCPSegmentation<PointT>
  {
      typedef PointXYZINormal WeightSACPointType;
      typedef LCCPSegmentation<PointT> LCCP;
      // LCCP typedefs
      using typename LCCP::EdgeID;
```

This is the line where the compiler chokes (`cpc_segmentation.hpp line 90`):
```
template <typename PointT> void
pcl::CPCSegmentation<PointT>::applyCuttingPlane (uint32_t depth_levels_left)
{
  typedef std::map<uint32_t, pcl::PointCloud<WeightSACPointType>::Ptr> SegLabel2ClusterMap;
  typedef std::map<uint32_t, std::vector<EdgeID> > SegLabel2Edges; <---- this one!
```

I would appreciate any feedback here. Cheers and thanks!

**Comment by taketwo** [comment]
No idea. Looks like perfectly valid C++ for me.

Did you try to spell out the LCCP type completely in the declaration: `using typename LCCPSegmentation<PointT>::EdgeID;` Would this make any difference?

If not we may try to just `typedef` the edge id instead of `using` it, i.e. `typedef typename LCCP::EdgeID EdgeID;`

**Comment by taketwo** [comment]
Regarding the `PointXYZINormal` problem, I think the approach you used to solve it is the best we can do. I would propose a minor cosmetic change though: unconditionally include `.h` files on top, and move the ugly `ifdef`s to the bottom, where it could be merged with already existing checks:

```cpp
#ifdef PCL_NO_PRECOMPILE
#include <pcl/segmentation/impl/cpc_segmentation.hpp>
#elif defined(PCL_ONLY_CORE_POINT_TYPES)
// pcl::PointXYZINormal is not a core point type (so we cannot use the precompiled classes here)
#include <pcl/sample_consensus/impl/sac_model_plane.hpp>
#include <pcl/segmentation/impl/extract_clusters.hpp>  
#endif
```

By the way this will slightly improve the inclusion logic (because in case of `NO_PRECOMPILE` the implementation files will be included anyway).

**Comment by mschoeler** [comment]
Good idea and thanks for the advice. `typedef` did the trick. Still strange that my system did not choke on that one (maybe a different version of gcc). Just added your cosmetic change, too.

**Comment by taketwo** [comment]
LGTM, I think it's ready to be merged.

**Pull request merged**